### ISSUE TITLE
How to keep session alive when using async websockets?

### ISSUE THREAD

**Issue by mllamazing** [aaugustin/websockets#161]
I have this code to interact with a websocket api using `async` and `websokets` python libraries:

```
#!/usr/bin/env python3

import sys, json
import asyncio
from websockets import connect

class AsyncWebsocket():
    async def __aenter__(self):
        self._conn = connect('wss://ws.myws.com/v2')
        self.websocket = await self._conn.__aenter__()        
        return self

    async def __aexit__(self, *args, **kwargs):
        await self._conn.__aexit__(*args, **kwargs)

    async def send(self, message):
        await self.websocket.send(message)

    async def receive(self):
        return await self.websocket.recv()

class mtest():
    def __init__(self, api_token):
        self.aws        = AsyncWebsocket()
        self.loop       = asyncio.get_event_loop()
        self.api_token  = api_token
        
        self.authorize()

    def authorize(self):
        jdata = self.__async_exec({
                                    'authorize': self.api_token
                                  })

        try:
            print (jdata['email'])
            ret = True
        except:
            ret = False

        return ret

    def sendtest(self):

        jdata = self.__async_exec({
                                    "hello": 1
                                  })

        print (jdata)


    def __async_exec(self, jdata):
        try:
            ret = json.loads(self.loop.run_until_complete(self.__async_send_recieve(jdata)))
        except:
            ret = None

        return ret

    async def __async_send_recieve(self, jdata):
        async with self.aws:
            await self.aws.send(json.dumps(jdata))
            return await self.aws.receive()
```

So I have the following in main.py:
    
```
from webclass import *

a = mtest('12341234')
print (a.sendtest())
```

The problem is that it doesn't preserve the authorized session, so this is the output:

```
root@ubupc1:/home/dinocob# python3 test.py 
asd@gmail.com
{'error': {'message': 'Please log in.', 'code': 'AuthorizationRequired'}}
```

As you see, the login call are working ok, but when calling and sending the `hello` in `sendtest` function, the session is not the same. 

 - Where is destroyed the session? 
 - How could I preserve it (without drastically
   modifying my class structure)?

**Comment by aaugustin** [comment]
I'm sorry, I'm not sure how I can help you.

You're going to great lengths to avoid asyncio's coroutine-based programming models, depriving yourself from its benefits — which you then miss.

Generally speaking classes don't work well with asyncio, you should maintain state in the local variables of your coroutines (or explicitly in shared state containers).

Write a corountine that opens the connection and calls other corountines to do whatever it needs to do. Then call `asyncio.get_event_loop().run_until_complete(your_coroutine)`. That must be the entrypoint of your whole program.

Your `__async_exec` method is actually synchronous, since it blocks until the event loop finishes, so I suspect you're still quite confused about how asyncio works.

Finally I have no idea how the server you're interacting with maintains sessions, making it hand to answer questions like "where is the session detroyed?"

----

To sum up:
- Your class structure may be appropriate for a callback-based programming model.
- The point of asyncio is to provide a higher-level, coroutine-based programming model.
- The point of websockets is to provide a coroutine-based API for websocket connections.
- You should either embrace the coroutine model or not use tools designed for this model.
